library/cwm/test/cwm_test.rb

```diff
@@ -0,0 +1,229 @@
+      expect(run_widgets[0]["init"].remote_method).to eq(method(:w1_init))
+      expect(run_widgets[0]["store"].remote_method).to eq(method(:generic_save))
+      expect(run_widgets[1]["init"].remote_method).to eq(method(:generic_init))
+      expect(run_widgets[1]["store"].remote_method).to eq(method(:w2_store))
+    end
+  end
+
+  # used by CWMTab and CWM::ReplacePoint
+  describe "#initWidgets" do
+    it "calls init methods" do
+      expect(self).to receive(:w1_init).with("w1")
+      expect(self).to receive(:generic_init).with("w2")
+      subject.initWidgets(run_widgets)
+    end
+
+    # used via GetProcessedWidget by yast2-slp-server and yast2
+    xit "sets @processed_widget" do
+    end
+
+    xit "sets ValidChars" do
+    end
+  end
+
+  # used by CWMTab and CWM::ReplacePoint
+  describe "#saveWidgets" do
+    it "calls store methods" do
+      expect(self).to receive(:generic_save).with("w1", "ID" => :event)
+      expect(self).to receive(:w2_store).with("w2", "ID" => :event)
+      subject.saveWidgets(run_widgets, "ID" => :event)
+    end
+
+    # used via GetProcessedWidget by yast2-slp-server and yast2
+    xit "sets @processed_widget" do
+    end
+  end
+
+  # used by CWMTab and CWM::ReplacePoint
+  describe "#handleWidgets" do
+    it "calls the handle methods" do
+      expect(self).to receive(:w1_handle).with("w1", "ID" => :event).and_return(nil)
+      expect(self).to receive(:w2_handle).with("w2", "ID" => :event).and_return(nil)
+      expect(subject.handleWidgets(run_widgets, "ID" => :event)).to eq(nil)
+    end
+
+    it "breaks the loop if a handler returns non-nil" do
+      expect(self).to receive(:w1_handle).with("w1", "ID" => :event).and_return(:foo)
+      expect(self).to_not receive(:w2_handle)
+      expect(subject.handleWidgets(run_widgets, "ID" => :event)).to eq(:foo)
+    end
+
+    xit "sets @processed_widget" do
+    end
+
+    xit "filters the events if 'handle_events' is specified" do
+    end
+  end
+
+  describe "#validateWidgets" do
+    it "calls the validate methods" do
+      expect(self).to receive(:w1_validate).with("w1", "ID" => :event).and_return(true)
+      expect(self).to receive(:w2_validate).with("w2", "ID" => :event).and_return(true)
+      expect(subject.validateWidgets(run_widgets, "ID" => :event)).to eq(true)
+    end
+
+    it "breaks the loop if a handler returns false" do
+      expect(self).to receive(:w1_validate).with("w1", "ID" => :event).and_return(false)
+      expect(self).to_not receive(:w2_validate)
+      expect(subject.validateWidgets(run_widgets, "ID" => :event)).to eq(false)
+    end
+
+    # SetValidationFailedHandler
+    xit "calls validation_failed_handler if..." do
+    end
+  end
+
+  describe "#PrepareDialog" do
+    it "returns early if the term is empty" do
+      expect(subject).to_not receive(:ProcessTerm)
+      expect(subject.PrepareDialog(VBox(), test_widgets)).to eq(VBox())
+    end
+
+    it "prepares args for ProcessTerm" do
```

Comment:
please test here that it does not modify stringterm parameter, which is quite important thing when we reduce deep_copies

Reply:
This is surprisingly hard to test in general, because Ruby does not have `deep_clone` or `deep_freeze`. Fortunately `deep_copy` helps :-)